### ISSUE TITLE
Getting gradient of element of tensor wrt the element itself

### ISSUE THREAD

**Issue by Microno95** [pytorch/pytorch#23941]
## 🐛 Bug

I have a tensor b, for which I am trying to get the gradient of an element of b wrt a sub-tensor of b. The results are unexpected...

## To Reproduce
```python
import torch
b = torch.zeros([2, 2], dtype=torch.float32, requires_grad=True)
print(torch.autograd.grad(b[0][0], b[0], allow_unused=True)
c = b[0]
print(torch.autograd.grad(c[0], c, allow_unused=True)
```

Running the script above, I would expect to obtain the same output for the two print statements, but I do not.

Instead I obtain `(None,)` for the first print and `(tensor([1., 0.], dtype=torch.float32),)` for the second print.

## Environment

The output from `collect_env.py`:
```
Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Arch Linux
GCC version: (GCC) 8.2.1 20181127
CMake version: version 3.13.3

Python version: 3.6
Is CUDA available: No
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti

Nvidia driver version: 415.27
cuDNN version: /usr/lib/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip3] numpy==1.16.0
[pip3] torch==0.4.1.post2
[pip3] torchvision==0.2.1
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.1.0            py36h516909a_0    conda-forge
[conda] mkl_fft                   1.0.10                   py36_0    conda-forge
[conda] mkl_random                1.0.2                    py36_0    conda-forge
[conda] pytorch                   1.1.0           py3.6_cuda10.0.130_cudnn7.5.1_0    pytorch
[conda] torchvision               0.3.0           py36_cu10.0.130_1    pytorch
```

(Is CUDA available should be Yes, but the script did not detect it correctly...)

**Comment by ezyang** [comment]
```
print(torch.autograd.grad(b[0][0], b[0], allow_unused=True)
```

This doesn't do what you want because the autograd graph for `b[0][0]` is not related to `b[0]` (no CSE happens here.) The second code is correct and you should do it that way.

**Comment by ezyang** [comment]
(I suppose you could argue that `b[0]` is a view and that's why they should be shared, but that's not how our implementation of views works today.)

**Comment by Microno95** [comment]
I see, that makes sense and I suspected as much. 

I assume that `torch.autograd.grad(b[0, 0], b[0], allow_unused=True)` also gives `(None,)` because `b[0, 0]` is implemented as `b[0][0]` under the hood?

**Comment by ezyang** [comment]
Well, it doesn't really matter how `b[0, 0]` is implemented under the hood: you didn't compute `b[0, 0]` as a function of the exact output of `b[0]` you passed in ;)

**Comment by Microno95** [comment]
Ohhh, I get it now. So essentially, the graph of `b[0,0]` is distinct from the graph of `b[0]` because `b[0,0]` was not computed from `b[0]`.
